cpu/esp8266/periph/spi.c

```diff
@@ -0,0 +1,320 @@
+                             spi_clkcnt_N = 4;      // 4 cycles results into 10 MHz
+                             break;
+        case SPI_CLK_5MHZ:   spi_clkdiv_pre = 2;    // predivides 80 MHz to 40 MHz
+                             spi_clkcnt_N = 8;      // 8 cycles results into 5 MHz
+                             break;
+        case SPI_CLK_1MHZ:   spi_clkdiv_pre = 2;    // predivides 80 MHz to 40 MHz
+                             spi_clkcnt_N = 40;     // 40 cycles results into 1 MHz
+                             break;
+        case SPI_CLK_400KHZ: spi_clkdiv_pre = 20;   // predivides 80 MHz to 4 MHz
+                             spi_clkcnt_N = 10;     // 10 cycles results into 400 kHz
+                             break;
+        case SPI_CLK_100KHZ: spi_clkdiv_pre = 20;   // predivides 80 MHz to 4 MHz
+                             spi_clkcnt_N = 40;     // 20 cycles results into 100 kHz
+                             break;
+        default: spi_clkdiv_pre = 20;   // predivides 80 MHz to 4 MHz
+                 spi_clkcnt_N = 40;     // 20 cycles results into 100 kHz
+    }
+
+    // register values are set to deviders-1
+    spi_clkdiv_pre--;
+    spi_clkcnt_N--;
+
+    DEBUG("%s spi_clkdiv_prev=%lu spi_clkcnt_N=%lu\n", __func__, spi_clkdiv_pre, spi_clkcnt_N);
+
+    // SPI clock is derived from system bus frequency and should not be affected by
+    // CPU clock
+
+    IOMUX.CONF &= ~(bus == 0 ? IOMUX_CONF_SPI0_CLOCK_EQU_SYS_CLOCK
+                             : IOMUX_CONF_SPI1_CLOCK_EQU_SYS_CLOCK);
+    SPI(bus).CLOCK  = VAL2FIELD_M (SPI_CLOCK_DIV_PRE, spi_clkdiv_pre) |
+                      VAL2FIELD_M (SPI_CLOCK_COUNT_NUM, spi_clkcnt_N) |
+                      VAL2FIELD_M (SPI_CLOCK_COUNT_HIGH, (spi_clkcnt_N+1)/2-1) |
+                      VAL2FIELD_M (SPI_CLOCK_COUNT_LOW, spi_clkcnt_N);
+
+    DEBUG("%s IOMUX.CONF=%08lx SPI(bus).CLOCK=%08lx\n",
+          __func__, IOMUX.CONF, SPI(bus).CLOCK);
+
+    return SPI_OK;
+}
+
+void spi_release(spi_t bus)
+{
+    // see spi_init
+    CHECK_PARAM (bus == SPI_DEV(0));
+
+    // release the bus
+    mutex_unlock(&_spi_lock[bus]);
+}
+
+/*
+ * Following functions are from the hardware SPI driver of the esp-open-rtos
+ * project.
+ *
+ * Copyright (c) Ruslan V. Uss, 2016
+ * BSD Licensed as described in the file LICENSE
+ * https://github.com/SuperHouse/esp-open-rtos/blob/master/LICENSE
+ */
+
+inline static void _set_size(uint8_t bus, uint8_t bytes)
+{
+    uint32_t bits = ((uint32_t)bytes << 3) - 1;
+    SPI(bus).USER1 = SET_FIELD(SPI(bus).USER1, SPI_USER1_MISO_BITLEN, bits);
+    SPI(bus).USER1 = SET_FIELD(SPI(bus).USER1, SPI_USER1_MOSI_BITLEN, bits);
+}
+
+inline static void _wait(uint8_t bus)
+{
+    while (SPI(bus).CMD & SPI_CMD_USR)
+        ;
+}
+
+inline static void _start(uint8_t bus)
+{
+    SPI(bus).CMD |= SPI_CMD_USR;
+}
+
+inline static void _store_data(uint8_t bus, const void *data, size_t len)
+{
+    uint8_t words = len / 4;
+    uint8_t tail = len % 4;
```

Comment:
since it is 4 any benifit in using 
`uint8_t words = len>>2;`
`uint8_t tail = len & 0x03;`

Reply:
This function is part of the code that was extracted from the SPI implementation in ```esp-open-rtos``` as stated in copyright notice at the beginning of this code. My intention was to let this part of the code as it is and not to modify it.

At the end the result is the same because of optimizations of the compiler. Integer arithmetic of powers of 2 are always mapped to shifts or AND / OR operations by the compiler.

Reply:
My bad, I missed that.  Compilers are pretty darn smart.